routing/road_access_serialization.hpp

```diff
@@ -26,36 +34,138 @@ class RoadAccessSerializer final
     uint32_t const header = kLatestVersion;
     WriteToSink(sink, header);
 
-    auto const & privateRoads = roadAccess.GetPrivateRoads();
-    ASSERT(std::is_sorted(privateRoads.begin(), privateRoads.end()), ());
-    WriteToSink(sink, base::checked_cast<uint32_t>(privateRoads.size()));
-    if (!privateRoads.empty())
-      WriteVarUint(sink, privateRoads[0]);
-    for (size_t i = 1; i < privateRoads.size(); ++i)
-    {
-      uint32_t const delta = privateRoads[i] - privateRoads[i - 1];
-      WriteVarUint(sink, delta);
-    }
+    for (auto const routerType : RoadAccess::GetSupportedRouterTypes())
+      SerializeOneRouterType(sink, roadAccess.GetTypes(routerType));
   }
 
   template <class Source>
   static void Deserialize(Source & src, RoadAccess & roadAccess)
   {
     uint32_t const header = ReadPrimitiveFromSource<uint32_t>(src);
     CHECK_EQUAL(header, kLatestVersion, ());
-    size_t numPrivateRoads = base::checked_cast<size_t>(ReadPrimitiveFromSource<uint32_t>(src));
-    std::vector<uint32_t> privateRoads(numPrivateRoads);
-    if (numPrivateRoads > 0)
-      privateRoads[0] = ReadVarUint<uint32_t>(src);
-    for (size_t i = 1; i < numPrivateRoads; ++i)
+
+    for (auto const routerType : RoadAccess::GetSupportedRouterTypes())
     {
-      uint32_t delta = ReadVarUint<uint32_t>(src);
-      privateRoads[i] = privateRoads[i - 1] + delta;
+      RoadAccessTypesMap m;
+      DeserializeOneRouterType(src, m);
+      roadAccess.SetTypes(routerType, std::move(m));
     }
-    roadAccess.SetPrivateRoads(move(privateRoads));
   }
 
 private:
+  template <typename Sink>
+  static void SerializeOneRouterType(Sink & sink, RoadAccessTypesMap const & m)
+  {
+    std::array<std::vector<Segment>, static_cast<size_t>(RoadAccess::Type::Count)>
+        segmentsByRoadAccessType;
+    for (auto const & kv : m)
+      segmentsByRoadAccessType[static_cast<size_t>(kv.second)].push_back(kv.first);
+
+    for (auto & segs : segmentsByRoadAccessType)
+    {
+      std::sort(segs.begin(), segs.end());
+      SerializeSegments(sink, segs);
+    }
+  }
+
+  template <typename Source>
+  static void DeserializeOneRouterType(Source & src, RoadAccessTypesMap & m)
+  {
+    m.clear();
+    for (size_t i = 0; i < static_cast<size_t>(RoadAccess::Type::Count); ++i)
+    {
+      std::vector<Segment> segs;
+      DeserializeSegments(src, segs);
+      for (auto const & seg : segs)
+        m[seg] = static_cast<RoadAccess::Type>(i);
+    }
+  }
+
+  // todo(@m) This code borrows heavily from traffic/traffic_info.hpp:SerializeTrafficKeys.
+  template <typename Sink>
+  static void SerializeSegments(Sink & sink, vector<Segment> const & segments)
```

Comment:
Generally speaking SerializeSegments() may be used in different context. What do you think about adding CHECK(std::is_sorted(segments....), ())?